server/storage/mvcc/revision.go

```diff
@@ -94,14 +97,25 @@ func BucketKeyToBytes(rev BucketKey, bytes []byte) []byte {
 	return bytes
 }
 
-func BytesToBucketKey(bytes []byte) BucketKey {
+func BytesToBucketKey(bytes []byte) (BucketKey, error) {
+	if (len(bytes) != revBytesLen) && (len(bytes) != markedRevBytesLen) {
+		return BucketKey{}, fmt.Errorf("invalid revision length: %d", len(bytes))
+	}
+	if bytes[8] != '_' {
+		return BucketKey{}, fmt.Errorf("invalid separator in bucket key: %q", bytes[8])
+	}
+	main := int64(binary.BigEndian.Uint64(bytes[0:8]))
+	sub := int64(binary.BigEndian.Uint64(bytes[9:]))
+	if main < -1 || sub < 0 { // finishedCompactRev can be -1
+		return BucketKey{}, errors.New("negative revision")
+	}
```

Comment:
Revision numbers must be positive. Negative values are invalid. I added these extra checks to be able to detect corruptions on database when checking snapshot status on disk.

Reply:
Suggest not to change the signature. Instead we can add assertion (panicking when it breaks any invariant properties)

Reply:
Also I see your comment "`// finishedCompactRev can be -1`". Can you elaborate in what exact case `finishedCompactRev` may be -1?

Reply:
It's initially set to one:
https://github.com/cenkalti/etcd/blob/0316c5af0e7e9dc02e206e8f26df9fc4c5bc2118/server/storage/mvcc/kvstore.go#L100

Reply:
I don't see any case etcd may persist the `-1`.

Reply:
Reverted back the signature change. Now it's panicking in case of invalid revision bytes.